clients/drcachesim/tools/filter/encoding_filter.h

```diff
@@ -0,0 +1,173 @@
+                          std::vector<trace_entry_t> &last_encoding) override
+    {
+        /* TODO i#6662: modify trace_entry_t header entry to regdeps ISA, instead of the
+         * real ISA of the incoming trace.
+         */
+
+        /* We have encoding to convert.
+         * Normally the sequence of trace_entry_t(s) looks like:
+         * [encoding,]+ instr_with_PC, [read | write]*
+         * (+ = one or more, * = zero or more)
+         * If we enter here, trace_entry_t is instr_with_PC.
+         */
+        if (is_any_instr_type(static_cast<trace_type_t>(entry.type)) &&
+            !last_encoding.empty()) {
+            /* Gather real ISA encoding bytes looping through all previously saved
+             * encoding bytes in last_encoding.
+             */
+            const app_pc pc = reinterpret_cast<app_pc>(entry.addr);
+            byte encoding[MAX_ENCODING_LENGTH];
+            memset(encoding, 0, sizeof(encoding));
+            uint encoding_offset = 0;
+            for (auto &trace_encoding : last_encoding) {
+                memcpy(encoding + encoding_offset, trace_encoding.encoding,
+                       trace_encoding.size);
+                encoding_offset += trace_encoding.size;
+            }
+
+            /* Genenerate the real ISA instr_t by decoding the encoding bytes.
+             */
+            instr_t instr;
+            instr_init(dcontext_.dcontext, &instr);
+            app_pc next_pc = decode_from_copy(dcontext_.dcontext, encoding, pc, &instr);
+            if (next_pc == NULL || !instr_valid(&instr)) {
+                instr_free(dcontext_.dcontext, &instr);
+                error_string_ =
+                    "Failed to decode instruction " + to_hex_string(entry.addr);
+                return false;
+            }
+
+            /* Convert the real ISA instr_t into a regdeps ISA instr_t.
+             */
+            instr_t instr_regdeps;
+            instr_init(dcontext_.dcontext, &instr_regdeps);
+            instr_convert_to_isa_regdeps(dcontext_.dcontext, &instr, &instr_regdeps);
+
+            /* Obtain regdeps ISA instr_t encoding bytes.
+             */
+            byte ALIGN_VAR(REGDEPS_ALIGN_BYTES)
+                encoding_regdeps[REGDEPS_MAX_ENCODING_LENGTH];
+            app_pc next_pc_regdeps =
+                instr_encode(dcontext_.dcontext, &instr_regdeps, encoding_regdeps);
+
+            /* Compute number of trace_entry_t to contain regdeps ISA encoding.
+             * Each trace_entry_t record can contain 8 byte encoding.
+             */
+            uint trace_entry_encoding_size = (uint)sizeof(entry.addr); /* == 8 */
+            uint regdeps_encoding_size = (uint)(next_pc_regdeps - encoding_regdeps);
+            uint num_regdeps_encoding_entries =
+                ALIGN_FORWARD(regdeps_encoding_size, trace_entry_encoding_size) /
+                trace_entry_encoding_size;
+            last_encoding.resize(num_regdeps_encoding_entries);
```

Comment:
record_filter.cpp:643 says "we don't support partial encoding removal": this PR needs to either add that support or update that comment as this resize could result in partial encoding removal if it shrinks the encoding record count, right?

Reply:
From my understanding that refers to cases where a filter removes the encodings of an instruction which will then be encountered later, and when we look for that instruction encoding we won't be able to find them.
In the encodings2regdeps filter we always have the encoding of instructions, even if they can be larger or smaller than the original.
However, I need to investigate this issue more. It's not entirely clear to me what that code is trying to achieve.

Reply:
Yes that does sound right: "partial encoding removal" means removing some encodings from part of the trace but leaving others in other parts, which gets hard to support with the chunking.

Reply:
Ok, in that case we don't need to add that support yet.
I clarified this in the two comments mentioning lack of support for partial encodings removal.